src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs

```diff
@@ -158,20 +159,50 @@ public static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProto
                     context = new SafeDeleteSslContext((credential as SafeFreeSslCredentials)!, sslAuthenticationOptions);
                 }
 
-                bool done = Interop.OpenSsl.DoSslHandshake(((SafeDeleteSslContext)context).SslContext, inputBuffer, out output, out outputSize);
+                SecurityStatusPalErrorCode errorCode = Interop.OpenSsl.DoSslHandshake(((SafeDeleteSslContext)context).SslContext, inputBuffer, out output, out outputSize);
+
+                if (errorCode == SecurityStatusPalErrorCode.CredentialsNeeded)
+                {
+                    if (sslAuthenticationOptions.CertSelectionDelegate != null)
+                    {
+                        X509Certificate2? remoteCert = null;
+                        string[] issuers = CertificateValidationPal.GetRequestCertificateAuthorities(context);
+                        try
+                        {
+                            remoteCert = CertificateValidationPal.GetRemoteCertificate(context);
+                            if (sslAuthenticationOptions.ClientCertificates == null)
+                            {
+                                sslAuthenticationOptions.ClientCertificates = new X509CertificateCollection();
+                            }
+                            X509Certificate2 clientCertificate = (X509Certificate2)sslAuthenticationOptions.CertSelectionDelegate(sslAuthenticationOptions.TargetHost!, sslAuthenticationOptions.ClientCertificates, remoteCert, issuers);
+                            if (clientCertificate != null && clientCertificate.HasPrivateKey)
+                            {
+                                sslAuthenticationOptions.CertificateContext = SslStreamCertificateContext.Create(clientCertificate);
+                            }
+                        }
+                        finally
+                        {
+                            remoteCert?.Dispose();
+                        }
+                    }
```

Comment:
except for line 180, this code is the same as in SslStreamPal.OSX.cs. Can it be extracted to a common helper function?

Reply:
maybe, but I don't know where that would live at this moment. We can look at it as follow-up.  I originally tried to make it like Windows, but I could not make it work. So there is another very similar copy in schannel.cs.